metriks/gauge.go

```diff
@@ -0,0 +1,88 @@
+package metriks
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/armon/go-metrics"
+)
+
+const (
+	defaultGaugeDuration = time.Second * 10
+)
+
+// PersistentGauge will report on an interval the value to the metrics collector.
+// Every call to the methods to modify the value immediately report, but if we
+// don't have a change inside the window (default 10s) after the last report
+// we will report the current value.
+type PersistentGauge struct {
+	name  string
+	value int32
+	tags  []metrics.Label
+
+	ticker *time.Ticker
+	cancel context.CancelFunc
+	dur    time.Duration
+}
+
+// Set will replace the value with a new one, it returns the old value
+func (g *PersistentGauge) Set(value int32) int32 {
+	v := atomic.SwapInt32(&g.value, value)
+	g.report(value)
+	return v
+}
+
+// Inc will +1 to the current value and return the new value
+func (g *PersistentGauge) Inc() int32 {
+	v := atomic.AddInt32(&g.value, 1)
+	g.report(v)
+	return v
+}
+
+// Dec will -1 to the current value and return the new value
+func (g *PersistentGauge) Dec() int32 {
+	v := atomic.AddInt32(&g.value, -1)
+	g.report(v)
+	return v
+}
+
+func (g *PersistentGauge) report(v int32) {
+	Gauge(g.name, float32(v), g.tags...)
+	g.ticker.Reset(g.dur)
+}
+
+func (g *PersistentGauge) start(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-g.ticker.C:
+			g.report(g.value)
+		}
```

Comment:
What do you think about using the ScheduledExecutor util here?

Reply:
I was thinking about that ~ one thing that I was seeing is that I couldn't do a 'Reset' on it. What I was intending on doing with this was that it would only report a value, if we hadn't reported in x seconds.

Reply:
Oh I get this now. So, maybe an alternative would be don't report on each operation, just record the value, then the timer would report the current value every 10secs evenly?

Reply:
I made the `ScheduledGauge` that would do this. But in even testing it I found some oddness in the sense of "when is it actually reported". I think having 2 of them could be nice, one that will internally track the value and the second that will use a callback